### PR TITLE
- add kuba-zip/0.1.31

### PR DESCRIPTION
**Pull request by SSE4**
Specify library name and version:  **zip/0.1.30**

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/zip/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.c)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
+
```
Comment (prince-chrismc) [inline recipes/zip/all/test_package/CMakeLists.txt:9]: It's also nice to test the official targets from upstream

```suggestion

find_package(zip REQUIRED)

add_executable(${PROJECT_NAME} test_package.c)
target_link_libraries(${PROJECT_NAME} zip::zip)
```

File: recipes/zip/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        cmake = CMake(self)
+        cmake.definitions["CMAKE_DISABLE_TESTING"] = True
+        cmake.configure()
```
Comment (prince-chrismc) [inline recipes/zip/all/conanfile.py:47]: Or we can remove https://github.com/conan-io/conan-center-index/pull/5081/files#diff-c5d00f02f2f77b30176b2669e7f269506265601816d1c0dc277900fdcf0408adR25-R27

```suggestion
        cmake.configure(build_folder=self._build_subfolder)
```

File: recipes/zip/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import os
+from conans import ConanFile, CMake, tools
+
+
+class ZipConan(ConanFile):
+    name = "zip"
```
Comment (Croydon) [inline recipes/zip/all/conanfile.py:6]: urgh. I really don't like that this lib is named after a file format which it did not invent

Comment (SSE4) [reply recipes/zip/all/conanfile.py:6]: neither me

File: recipes/zip/all/conanfile.py (hunk 2 above)
Comment (SSE4) [reply recipes/zip/all/conanfile.py:47]: well, not sure if it improves anything, doesn't seem to be required (builds just fine without this line), but only makes code longer -> harder to read and understand

File: recipes/zip/all/test_package/CMakeLists.txt (hunk 1 above)
Comment (SSE4) [reply recipes/zip/all/test_package/CMakeLists.txt:9]: same:
well, not sure if it improves anything, doesn't seem to be required (builds just fine without this line), but only makes code longer -> harder to read and understand

File: recipes/zip/all/conanfile.py (hunk 3 above)
Comment (prince-chrismc) [reply recipes/zip/all/conanfile.py:6]: We have our `author-repo` name which could be applied?

Comment (Croydon) [reply recipes/zip/all/conanfile.py:6]: Agreed. We applied the `authorname-projectname` several times in the past when we were thinking that the name is too generic

Comment (uilianries) [reply recipes/zip/all/conanfile.py:6]: Yes! `kuba-zip` please

File: recipes/zip/all/conanfile.py (hunk 2 above)
Comment (uilianries) [reply recipes/zip/all/conanfile.py:47]: It helps for development, separated build cache from copied sources.

File: recipes/zip/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
```
Comment (prince-chrismc) [inline recipes/zip/all/conanfile.py:28]: This is not used

```suggestion

```

File: recipes/zip/all/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake", "cmake_find_package"
```
Comment (prince-chrismc) [inline recipes/zip/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake"
```

**Review by prince-chrismc — COMMENTED** [review]
There's some cleanup we can do if the other suggestions are not applied

File: recipes/zip/all/conanfile.py (hunk 3 above)
Comment (SSE4) [reply recipes/zip/all/conanfile.py]: renamed

File: recipes/kuba-zip/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
```
Comment (SSE4) [inline recipes/kuba-zip/all/conanfile.py:28]:
```suggestion
```

File: recipes/kuba-zip/all/test_package/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake", "cmake_find_package"
```
Comment (SSE4) [inline recipes/kuba-zip/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake"
```

File: recipes/zip/all/conanfile.py (hunk 2 above)
Comment (SSE4) [reply recipes/zip/all/conanfile.py]: maybe it does, but it fails to build such way for me, so please don't suggest it, it doesn't work

**Review by uilianries — DISMISSED** [review]

**Review by prince-chrismc — DISMISSED** [review]

**Comment by SpaceIm** [comment]
fPIC option is not honored: https://github.com/kuba--/zip/blob/f72dce22bd25d10f8575ec54cf51bf5a5fdcb5ba/CMakeLists.txt#L16

**Comment by SpaceIm** [comment]
This library has miniz headers in its source code. Could it conflict with `miniz` recipe?

**Comment by SSE4** [comment]
> fPIC option is not honored: https://github.com/kuba--/zip/blob/f72dce22bd25d10f8575ec54cf51bf5a5fdcb5ba/CMakeLists.txt#L16

fPIC option is honored:
https://github.com/conan-io/conan-center-index/pull/5081/files#diff-700faebf6f544847d0c88a9b2c17976e63aab94f3028908bebf80f68c3e482a0R35

also, fPIC hook passes

**Comment by SpaceIm** [comment]
> 
> 
> > fPIC option is not honored: https://github.com/kuba--/zip/blob/f72dce22bd25d10f8575ec54cf51bf5a5fdcb5ba/CMakeLists.txt#L16
> 
> fPIC option is honored:
> https://github.com/conan-io/conan-center-index/pull/5081/files#diff-700faebf6f544847d0c88a9b2c17976e63aab94f3028908bebf80f68c3e482a0R35
> 
> also, fPIC hook passes

fPIC hooks doesn't check that static lib is really PIC or not. `fPIC=False` is not honored due to upstream CMakeLists.

**Comment by SSE4** [comment]
> This library has miniz headers in its source code. Could it conflict with `miniz` recipe?

from their readme, it looks like a fork of original miniz with hacked functions + more wrappers.
we may consider adding `provides = "miniz"`, if it's needed.

**Comment by SSE4** [comment]
> fPIC hook doesn't check that static lib is really PIC or not. `fPIC=False` is not honored due to upstream CMakeLists.

if so, what can we do? this is going beyond our control, probably we can only raise `ConanInvalidConfiguration` for `fPIC=False` if it's not supported.

**Comment by SpaceIm** [comment]
patch or tools.replace_in_file

**Comment by SSE4** [comment]
> patch or tools.replace_in_file

I don't like that, to be honest. if upstream forces fPIC=True, they might have a valid reasons to do that, and trying to workaround that may lead into some issues. I'd prefer to just reject the configurations upstream doesn't support.

File: recipes/kuba-zip/all/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.4)
+project(cmake_wrapper)
+
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
+conan_basic_setup()
+
+set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (SpaceIm) [inline recipes/kuba-zip/all/CMakeLists.txt:7]: OK.

Surprinsingly, upstream CMakeLists calls generate_export_headers, but don't use generated file 😩

Seems like author tried something, then mixed different solutions: https://github.com/kuba--/zip/pull/122

Comment (SSE4) [reply recipes/kuba-zip/all/CMakeLists.txt:7]: yeah, maybe author is not so familiar with CMake

**Comment by SpaceIm** [comment]
> 
> 
> > patch or tools.replace_in_file
> 
> I don't like that, to be honest. if upstream forces fPIC=True, they might have a valid reasons to do that, and trying to workaround that may lead into some issues. I'd prefer to just reject the configurations upstream doesn't support.

Pretty sure that author has not valid reason to force PIC, except to an attempt to build both shared and static in an old PR maybe: https://github.com/kuba--/zip/pull/122

File: recipes/kuba-zip/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("UNLICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
```
Comment (SpaceIm) [inline recipes/kuba-zip/all/conanfile.py:57]:
```suggestion
    def package_info(self):
         self.cpp_info.names["cmake_find_package"] = "zip"
         self.cpp_info.names["cmake_find_package_multi"] = "zip"
```

**Comment by SpaceIm** [comment]
PIC issue solved: https://github.com/kuba--/zip/pull/177

File: recipes/kuba-zip/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        cmake = CMake(self)
+        cmake.definitions["CMAKE_DISABLE_TESTING"] = True
```
Comment (SSE4) [inline recipes/kuba-zip/all/conanfile.py:42]:
```suggestion
        cmake.definitions["CMAKE_DISABLE_TESTING"] = True
        cmake.definitions["ZIP_STATIC_PIC"] = self.options.get_safe("fPIC", False)
```

File: recipes/kuba-zip/all/CMakeLists.txt (hunk 8 above)
Comment (SpaceIm) [reply recipes/kuba-zip/all/CMakeLists.txt:7]: I've submitted https://github.com/kuba--/zip/pull/178

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/kuba-zip/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        cmake.definitions["ZIP_STATIC_PIC"] = self.options.get_safe("fPIC", False)
```
Comment (SpaceIm) [inline recipes/kuba-zip/all/conanfile.py:43]: Not sure to understand, it's not in 0.1.30, a patch is required, or wait the next release.

Comment (mathbunnyru) [inline recipes/kuba-zip/all/conanfile.py:43]: Could we ask the upstream for a new release?
This way we will remove small windows change in our cmake and `ZIP_STATIC_PIC` will work properly.

**Review by mathbunnyru — CHANGES_REQUESTED** [review]

File: recipes/kuba-zip/all/conanfile.py (hunk 11 above)
Comment (prince-chrismc) [reply recipes/kuba-zip/all/conanfile.py:43]: It's deleted on Linux to assume always true =)

```suggestion
        cmake.definitions["ZIP_STATIC_PIC"] = self.options.get_safe("fPIC", True)
```

**Comment by SpaceIm** [comment]
new release: https://github.com/kuba--/zip/releases/tag/v0.1.31, thanks @jgsogo for asking

**Review by jgsogo — CHANGES_REQUESTED** [review]
Please, use release `0.1.31` and we can simplify the recipe a little bit 🚀

**Comment by SpaceIm** [comment]
Please consider https://github.com/SSE4/conan-center-index/pull/9

**Review by SpaceIm — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Please update the title =)

**Comment by SpaceIm** [comment]
Bottleneck in CI? It takes few seconds to build each configuration of this library.

**Comment by jgsogo** [comment]
> Bottleneck in CI? It takes few seconds to build each configuration of this library.

Yes, today we have running (among others) `boost`, `opencv`, `qt`, `icu` and our own Thulsa Doom which is `gdal` 😓 . Just be patient, we will get there.

**Comment by prince-chrismc** [comment]
CI failed but there was no message from the bot

**Pull request merged**